src/main/webapp/script.js

```diff
@@ -149,39 +159,229 @@ class AutocompleteDirectionsHandler {
+                } else if (!dataGrid[mapRow][mapCol]) {
+                    console.log("Step aqi dosnt exist because mapCol doesn't exist in mapRow. So cant find and set stepAQI");
+                } else {
+                    stepAQI = dataGrid[mapRow][mapCol];
+                }
+
+                console.log(stepAQI);
+                totalValue += stepWeight * stepAQI;
+                totalWeight += stepWeight;
+                console.log("So far total value is: " + totalValue + " total weight is: "+ totalWeight);
+            }
+
+            // Counting the AQI at the end point of the leg as part of the score. 
+            // Set duration stepWeight as 60 seconds to account of parking upon arrival at end point. 
+            // TODO: Rosanna to rethink on this logic, would it stll apply with waypoints where there are a lot of legs? It can count for stopover time. 
+            let legWeight = 60;
+            let legEndlat = leg["end_location"].lat();
+            let legEndLng = leg["end_location"].lng();
+
+            let endPtRowCol = getGridIndex(legEndlat, legEndLng, aqDataPointsPerDegree);
+            let mapRow = endPtRowCol.row;
+            let mapCol = endPtRowCol.col;
+            let endptAQI = 0;
+
+            // Print to see in console whether mapRow or mapCol is undefined. 
+            if (!dataGrid[mapRow]) {
+                console.log("Row map doesnt exist and is undefined.");
+            } else if (!dataGrid[mapRow][mapCol]) {
+                console.log("Step aqi dosnt exist because mapCol doesn't exist in mapRow. So cant find and set stepAQI");
+            } else {
+                endptAQI = dataGrid[mapRow][mapCol];
+            }
+
+            console.log(endptAQI);
+            totalValue += legWeight * endptAQI;
+            totalWeight += legWeight;
+            console.log("So far total value is: " + totalValue + " total weight is: "+ totalWeight);
+        }
+
+        // Calculate total route Score 
+        let routeScore = totalValue / totalWeight; 
+        console.log("Total value is: " + totalValue + " totalweight is: " + totalWeight);
+        console.log("The Route Score is: " + routeScore);
+        return routeScore;
+    }
+
     calculateAndDisplayRoute() {
+        // If PlaceIDs are not placed in yet, dont calculate a route yet. 
         if (!this.originPlaceId || !this.destinationPlaceId) {
             return;
         }
+        const waypts = [];
+        const wayptAutocompleteArray = this.waypointAutocompleteArray;
+
+        for (const autocompleteObject of wayptAutocompleteArray) {
+            waypts.push ({
+                location: autocompleteObject.getPlace().name,
+                stopover: true,
+            })
+        }
+   
+        // Saving the class "this" as "me", as the definition of "this."  
+        // is changed in the reponse section and does not refer to the AutocompleteDirectionsHandler Class.
```

Comment:
This looks error-prone to me, could you pls also have a look? @sbland-g

Reply:
@Techget  @sbland-g  any suggestions on how else I can write this? Here I was mostly saving "this" that points to the handlerclass so i can refer back to it, as referecning "this" in this current block does not refer back to that class anymore, but it is still within that class

Reply:
maybe there should e a better variable name than "me"

Reply:
As mentioned in the comment; this is a weirdness of how classes and the 'this' reference works in JS. When the callback is invoked from within the the DirectionsService, it'll resolve 'this' to be referring the the DirectionsService instead of the context it was originally written in (here). In order to refer back to the original 'this' context here, we need to create an alias to it so it's available.

I'm not aware of a neater way to do it, and whilst a different variable name might be clearer ('originalThis'?), I think the use of 'me' us fine.